# Appendix II. Databases & Servers

## 0\) Genome Browsers

* UCSD Genome Browser [@youtube](https://youtu.be/eTgEtfI65hA) [@bilibili](https://player.bilibili.com/player.html?aid=30448417&cid=53132461&page=1)
* IGV [@youtube](https://youtu.be/6_1ZcVw7ptU) [@bilibili](https://player.bilibili.com/player.html?aid=30448472&cid=53133093&page=1)

## 1\) Projects

* **ENCODE**

  **Encyclopedia of DNA Elements**: a comprehensive parts list of functional elements in the human genome. [https://www.encodeproject.org/](https://www.encodeproject.org/)

* **1000 Genomes**

  most genetic variants with frequencies of at least 1% in the populations studied. [http://www.internationalgenome.org/](http://www.internationalgenome.org/)

* **TCGA**

  **The Cancer Genome Atlas:** From the analysis of over 11,000 tumors from 33 of the most prevalent forms of cancer, the Pan-Cancer Atlas provides a uniquely comprehensive, in-depth, and interconnected understanding of how, where, and why tumors arise in humans. As a singular and unified point of reference, the Pan-Cancer Atlas is an essential resource for the development of new treatments in the pursuit of precision medicine.

  * [portal](https://portal.gdc.cancer.gov/) 
  * [Cell papers](http://www.cell.com/pb-assets/consortium/pancanceratlas/pancani3/index.html)
  * [TCGA Research Network](http://cancergenome.nih.gov/)
  * [TCGA data portal](https://portal.gdc.cancer.gov/)
  * [Pan-TCGA](https://www.cell.com/pb-assets/consortium/pancanceratlas/pancani3/index.html)
  * [Broad GDAC Firehose](http://gdac.broadinstitute.org/)
  * [Broad GDAC FireBrowse](http://firebrowse.org/)
  * [TCGA/GTEx: Schultz Lab](https://www.nature.com/articles/sdata201861)
  * [Normalized data](https://figshare.com/articles/Data_record_3/5330593)
  * [TCGA/GTEx: GEPIA 2](http://gepia2.cancer-pku.cn/#dataset) 

* **CCLE** \(cancer cell lines\)

  **Cancer Cell Line Encyclopedia**

  * [CCLE](https://portals.broadinstitute.org/ccle)
  * CCLE RNA-seq BAM files：[https://portal.gdc.cancer.gov/legacy-archive/search/f?filters={"op":"and","content":\[{"op":"in","content":{"field":"cases.project.program.name","value":\["CCLE"\]}}\]}](https://portal.gdc.cancer.gov/legacy-archive/search/f?filters={"op":"and","content":[{"op":"in","content":{"field":"cases.project.program.name","value":["CCLE"]}}]})
  * [CCLE Raw data](https://www.ncbi.nlm.nih.gov/sra?term=PRJNA523380)
  * [CCLE Seven Bridges](https://www.sevenbridges.com/ccle/)

* **ICGC**
  * [ICGC](https://dcc.icgc.org/)
  * [PCAWG](https://docs.icgc.org/pcawg/data/)
* **Other Integrative Projects**
  * [COSMIC](https://cancer.sanger.ac.uk/cosmic)
  * [UCSC Xena](http://xena.ucsc.edu/)
  * [cBioPortal](http://www.cbioportal.org/)
  * [MSK-IMPACT](https://www.cbioportal.org/study/summary?id=msk_impact_2017)
  * [OncoKB](http://oncokb.org/)

## 2\) Databases

### \(1\) General

1. NCBI

   contribute to the NIH mission of ‘uncovering new knowledge’. [https://www.ncbi.nlm.nih.gov/](https://www.ncbi.nlm.nih.gov/)

2. UCSC

   genome browser for vertebrate. [https://genome.ucsc.edu/](https://genome.ucsc.edu/)

3. Ensemble

   genome annotation. [http://www.ensembl.org/index.html](http://www.ensembl.org/index.html)

### \(2\) Expression Data

1. GTEx

   gene expression in different tissues. [https://www.gtexportal.org/home/](https://www.gtexportal.org/home/)

2. Expression Atlas

   exploring gene expression results across species under different biological conditions. [https://www.ebi.ac.uk/gxa/home](https://www.ebi.ac.uk/gxa/home)

3. GEPIA

   gene expression in different TCGA tumor types. [http://gepia.cancer-pku.cn/index.html](http://gepia.cancer-pku.cn/index.html)

4. cBioPortal

   visualization, analysis and download of large-scale cancer genomics data sets. [http://www.cbioportal.org/index.do](http://www.cbioportal.org/index.do)

5. TCGA ncRNA

   [http://ibl.mdanderson.org/tanric/\_design/basic/index.html](http://ibl.mdanderson.org/tanric/_design/basic/index.html)​

### \(3\) Exosomal database

1.ExoCarta

A web-based compendium of exosomal cargo.[http://www.exocarta.org/](http://www.exocarta.org/)

2.exoRBase

A repository of circular RNA \(circRNA\), long non-coding RNA \(lncRNA\) and messenger RNA \(mRNA\) derived from RNA-seq data analyses of human blood exosomes. Experimental validations from published literature are also included.[http://www.exorbase.org/](http://www.exorbase.org/)

3.EVmiRNA

A database of miRNA profiling in extracellular vesicles. [http://bioinfo.life.hust.edu.cn/EVmiRNA](http://bioinfo.life.hust.edu.cn/EVmiRNA)

4.EV-TRACK

Transparent reporting and centralizing knowledge in extracellular vesicle research.[http://www.evtrack.org/](http://www.evtrack.org/)

5.EVpedia

An integrated database of high-throughput data for systemic analyses of extracellular vesicles. A community web portal for extracellular vesicles research. [http://evpedia.info](http://evpedia.info)

6.Vesiclepedia

A compendium for extracellular vesicles with continuous community annotation. PLoS Biology. 12, e1001450. [http://www.microvesicles.org/](http://www.microvesicles.org/)

7.EMBL-EBI

The EMBL-EBI exosomes database is a comprehensive annotated database of GO function analysis of human exosomal proteins. The database works closely with laboratory researchers \(Dr S Mathivanan - La Trobe Institute, Dr A Kalea - UCL\) and the exosomal database Exocarta \([http://www.exocarta.org/](http://www.exocarta.org/)\) and Vesiclepedia to identify annotations for focus Annotation of the target protein as well as the experimental methodology of the exosomes.[http://www.ebi.ac.uk/GOA/exosome](http://www.ebi.ac.uk/GOA/exosome)

8.Urinary Exosome Protein Database

This database of urinary exosome proteins is based on published protein mass spectrometry data from the NHLBI Epithelial Systems Biology Laboratory \(ESBL\). All data are from urinary exosomes isolated from healthy human volunteers.[https://hpcwebapps.cit.nih.gov/ESBL/Database/Exosome/](https://hpcwebapps.cit.nih.gov/ESBL/Database/Exosome/)

## 3\) Servers

### \(1\) General

1. BLAST

   finds regions of similarity between biological sequences. [https://blast.ncbi.nlm.nih.gov/Blast.cgi](https://blast.ncbi.nlm.nih.gov/Blast.cgi)

### \(2\) Gene functional analysis

1. David

   gene functional classification tool. [https://david.ncifcrf.gov/](https://david.ncifcrf.gov/)

2. KAAS

   KEGG Automatic Annotation Server. [https://www.genome.jp/kegg/kaas/](https://www.genome.jp/kegg/kaas/)

3. agriGO

   gene ontology \(GO\) enrichment analyses of plant and agricultural species. [http://systemsbiology.cau.edu.cn/agriGOv2/](http://systemsbiology.cau.edu.cn/agriGOv2/)

### \(3\) RNA secondary structure

1. RNAfold

   predict secondary structures of single stranded RNA or DNA sequences. [http://rna.tbi.univie.ac.at/cgi-bin/RNAWebSuite/RNAfold.cgi](http://rna.tbi.univie.ac.at/cgi-bin/RNAWebSuite/RNAfold.cgi)

2. RNAstructure

   RNA secondary structure prediction and analysis. [http://rna.urmc.rochester.edu/RNAstructureWeb/](http://rna.urmc.rochester.edu/RNAstructureWeb/)

### \(4\) miRNA targets prediction

1. TargetScan

   search for predicted microRNA targets in mammals [http://www.targetscan.org/vert\_72/](http://www.targetscan.org/vert_72/)

2. psRNATarget

   plant small RNA target analysis [https://plantgrn.noble.org/psRNATarget/](https://plantgrn.noble.org/psRNATarget/)

### \(5\) motif analysis

**sequence**

1. MEME motif based sequence analysis tools [http://meme-suite.org/](http://meme-suite.org/)
2. HOMER Software for motif discovery and next-gen sequencing analysis [http://homer.ucsd.edu/homer/motif/](http://homer.ucsd.edu/homer/motif/)

**structure**

1. RNApromo Computational prediction of RNA structural motifs involved in post transcriptional regulatory processes [https://genie.weizmann.ac.il/pubs/rnamotifs08/](https://genie.weizmann.ac.il/pubs/rnamotifs08/)
2. GraphProt modeling binding preferences of RNA-binding proteins [http://www.bioinf.uni-freiburg.de/Software/GraphProt/](http://www.bioinf.uni-freiburg.de/Software/GraphProt/)

